addons/account/models/account_move.py

```diff
@@ -1609,6 +1612,76 @@ def _compute_linked_attachment_id(self, attachment_field, binary_field):
         pass
 
+    @api.depends('partner_id', 'invoice_date', 'amount_total')
+    def _compute_abnormal_warning(self):
+        draft_invoices = self.filtered(lambda m: m.is_invoice() and m.state == 'draft' and m.amount_total)
+        other_moves = self - draft_invoices
+        other_moves.abnormal_amount_warning = False
+        other_moves.abnormal_date_warning = False
+        if draft_invoices:
+            draft_invoices.flush_recordset(['invoice_date', 'amount_total', 'partner_id', 'move_type', 'company_id'])
+            self.env.cr.execute("""
+                WITH previous_invoices AS (
+                      SELECT this.id,
+                             other.invoice_date,
+                             other.amount_total,
+                             LAG(other.invoice_date) OVER invoice - other.invoice_date AS date_diff
+                        FROM account_move this
+                        JOIN account_move other USING (partner_id, move_type, company_id, currency_id)
+                       WHERE other.state = 'posted'
+                         AND other.invoice_date <= this.invoice_date
+                         AND this.id = ANY(%s)
+                         AND this.id != other.id
+                      WINDOW invoice AS (PARTITION BY this.id ORDER BY other.invoice_date DESC)
+                ), stats AS (
+                      SELECT id,
+                             MAX(invoice_date)                OVER invoice AS last_invoice_date,
+                             AVG(date_diff)                   OVER invoice AS date_diff_mean,
+                             STDDEV_SAMP(date_diff)           OVER invoice AS date_diff_deviation,
+                             AVG(amount_total)         OVER invoice AS amount_mean,
+                             STDDEV_SAMP(amount_total) OVER invoice AS amount_deviation,
+                             ROW_NUMBER()                     OVER invoice AS row_number
+                        FROM previous_invoices
+                      WINDOW invoice AS (PARTITION BY id ORDER BY invoice_date DESC)
+                )
+                  SELECT id, last_invoice_date, date_diff_mean, date_diff_deviation, amount_mean, amount_deviation
+                    FROM stats
+                   WHERE row_number BETWEEN 10 AND 30
+                ORDER BY row_number ASC
+            """, [draft_invoices.ids])
+            result = {invoice: vals for invoice, *vals in self.env.cr.fetchall()}
+            for move in draft_invoices:
+                invoice_date = move.invoice_date or fields.Date.context_today(self)
+                (
+                    last_invoice_date, date_diff_mean, date_diff_deviation,
+                    amount_mean, amount_deviation,
+                ) = result.get(move._origin.id, (invoice_date, 0, 10000000000, 0, 10000000000))
+
+                if date_diff_mean > 25:
+                    date_diff_deviation += 1  # correct for varying days per month and leap years
```

Comment:
If we have a recurring invoice every month, the mean will be ~30.5 days, and the deviation ~1 day.
We need to add some wiggle room for the month of February otherwise it will trigger because 28 days is outside of the range

Reply:
Ok, thanks for the explanation: why not adding that example as comment? ;)

Reply:
Because I hate comments, I believe that code should be self explanatory as much as possible. In this case it is a little bit tricky so I hoped that one small comment was enough.

I added the full explanation.